pkg/quantile/sparse.go

```diff
@@ -0,0 +1,163 @@
+package quantile
+
+import (
+	"math"
+	"strings"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/quantile/summary"
+)
+
+// A Sketch for tracking quantiles
+type Sketch struct {
+	sparseStore
+
+	Basic summary.Summary
+}
+
+func (s *Sketch) String() string {
+	var b strings.Builder
+	printSketch(&b, s, Default())
+	return b.String()
+}
+
+// MemSize returns memory use in bytes:
+//   used: uses len(bins)
+//   allocated: uses cap(bins)
+func (s *Sketch) MemSize() (used, allocated int) {
+	const (
+		basicSize = int(unsafe.Sizeof(summary.Summary{}))
+	)
+
+	used, allocated = s.sparseStore.MemSize()
+	used += basicSize
+	allocated += basicSize
+	return
+}
+
+// InsertMany values into the sketch.
+func (s *Sketch) InsertMany(c *Config, values []float64) {
+	keys := getKeyList()
+	keys = keys[:0]
+
+	for _, v := range values {
+		s.Basic.Insert(v)
+		keys = append(keys, c.key(v))
+	}
+
+	s.insert(c, keys)
+	putKeyList(keys)
+}
+
+// Reset sketch to its empty state.
+func (s *Sketch) Reset() {
+	s.Basic.Reset()
+	s.count = 0
+	s.bins = s.bins[:0] // TODO: just release to a size tiered pool.
+}
+
+// Insert a single value into the sketch.
+// NOTE: InsertMany is much more efficient.
+func (s *Sketch) Insert(c *Config, vals ...float64) {
+	// TODO: remove this
+	s.InsertMany(c, vals)
+}
+
+// Merge o into s, without mutating o.
+func (s *Sketch) Merge(c *Config, o *Sketch) {
+	s.Basic.Merge(o.Basic)
+	s.merge(c, &o.sparseStore)
+}
+
+// Quantile returns v such that s.count*q items are <= v.
+//
+// Special cases are:
+//	Quantile(c, q <= 0)  = min
+//  Quantile(c, q >= 1)  = max
+func (s *Sketch) Quantile(c *Config, q float64) float64 {
+	switch {
+	case s.count == 0:
+		return 0
+	case q <= 0:
+		return s.Basic.Min
+	case q >= 1:
+		return s.Basic.Max
+	}
+
+	var (
+		n     float64
+		rWant = rank(s.count, q)
+	)
+
+	for i, b := range s.bins {
+		n += float64(b.n)
+		if n <= rWant {
+			continue
+		}
+
+		weight := (n - rWant) / float64(b.n)
+
+		vLow := c.f64(b.k)
+		vHigh := vLow * c.gamma.v
+
+		switch i {
+		case s.bins.Len():
+			vHigh = s.Basic.Max
+		case 0:
+			vLow = s.Basic.Min
+		}
+
+		// TODO|PROD: Interpolate between bucket boundaries, correctly handling min, max,
+		// negative numbers.
+		// with a gamma of 1.02, interpolating to the center gives us a 1% abs
+		// error bound.
+		return (vLow*weight + vHigh*(1-weight))
```

Comment:
Is this the best way to interpolate. For example, if there is one value in the bucket, then I think we should return the middle value of the bucket. And in that case, `n` is equal to `rWant+1` (because `n` was incremented by 1 in the last step), so we would return the lowest value of the bucket.
A better way to interpolate would be to use
```
weight := (n - rWant - 0.5) / float64(b.n)
```
We split the bucket in `b.n` sub-buckets with same width, we consider the sub-bucket of if 0-index `n - rWant - 1` and we return its middle value (`+ 0.5`).

Reply:
Yea this part is still in progress, I might remove the `Quantile` part from this PR (the agent doesn't need to query them).

That makes way more sense, I'll test this out for the server side version